### PR TITLE
[FLINK-13051][runtime] Replace the non-selectable stream task with the input-selectable one

### PR DESCRIPTION
**Pull request by sunhaibotb**
## What is the purpose of the change

This pull request replaces the non-selectable `StreamTask`/`InputProcessor` with the input-selectable ones. The aim is to simplify the Flink code and make it more efficient.

## Brief change log

  - Replaces the non-selectable stream task with the input-selectable one.
  - Drops the non-selectable StreamTask and InputProcessor.
  - Renames `TwoInputSelectableStreamTask` and `StreamTwoInputSelectableProcessor` to `TwoInputStreamTask` and `StreamTwoInputProcessor`respectively.

## Verifying this change

This change is already covered by existing tests, such as `TwoInputStreamTaskTest`, `StreamTaskCancellationBarrierTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java [hunk 1]
```diff
@@ -280,11 +286,17 @@ else if (recordOrMark.isWatermark()) {
 		}
 	}
 
+	private void updateInputSelection() {
+		if (inputSelector != null) {
+			inputSelection = inputSelector.nextSelection();
+		}
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java]: maybe add an else branch with 

```
else {
  inputSelection == InputSelection.ALL;
}
```
and re-use this in `prepareForProcessing`?

I don't think that would any measurable performance overhead, while would reduce code duplication by a bit?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/TwoInputStreamTaskTest.java [hunk 2]
```diff
@@ -623,36 +597,19 @@ public void testHandlingEndOfInput() throws Exception {
 		testHarness.invoke();
 		testHarness.waitForTaskRunning();
 
-		TestBoundedTwoInputOperator headOperator = (TestBoundedTwoInputOperator) testHarness.getTask().headOperator;
-
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/TwoInputStreamTaskTest.java]: Is this test simplification related to the rest of the commit? Was this test failing in the original form?

**Comment by pnowojski** [comment]
Thanks for the contribution @sunhaibotb. I really liked how did you structure the PR, it was easy to review/understand :)

Left two minor questions. I think the biggest question is performance. Have you measured it (modulo the unstable tests issue :( )?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/TwoInputStreamTaskTest.java (hunk 2 above)
Comment (sunhaibotb) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/TwoInputStreamTaskTest.java]: > Was this test failing in the original form?

**Comment by sunhaibotb** [comment]
Thanks for reviewing @pnowojski . The conflicts have resolved and test passed. 

The benchmarking show that the **twoInputMapSink** test is faster than the master branch, and the **twoInputOneIdleMapSink** test is slower than the master branch, but the maximum result of the **twoInputOneIdleMapSink** test is larger than the master branch. On the whole, I don't think there's regression in performance. What do you think? @pnowojski 

The test results of TwoInputBenchmark with bumping the number of forks to 15, are as follows.

- **Results of the branch: master**
```
Result "org.apache.flink.benchmark.TwoInputBenchmark.twoInputMapSink":
  8037.607 ±(99.9%) 415.634 ops/ms [Average]
  (min, avg, max) = (4630.272, 8037.607, 10680.036), stdev = 1516.361
  CI (99.9%): [7621.974, 8453.241] (assumes normal distribution)


Result "org.apache.flink.benchmark.TwoInputBenchmark.twoInputOneIdleMapSink":
  4823.837 ±(99.9%) 177.680 ops/ms [Average]
  (min, avg, max) = (3131.426, 4823.837, 5814.206), stdev = 648.230
  CI (99.9%): [4646.157, 5001.517] (assumes normal distribution)


# Run complete. Total time: 00:34:33

Benchmark                                  Mode  Cnt     Score     Error   Units
TwoInputBenchmark.twoInputMapSink         thrpt  150  8037.607 ± 415.634  ops/ms
TwoInputBenchmark.twoInputOneIdleMapSink  thrpt  150  4823.837 ± 177.680  ops/ms
```

- **Results of the branch: master + this PR**
```
Result "org.apache.flink.benchmark.TwoInputBenchmark.twoInputMapSink":
  9520.452 ±(99.9%) 218.321 ops/ms [Average]
  (min, avg, max) = (6911.392, 9520.452, 11013.918), stdev = 796.504
  CI (99.9%): [9302.130, 9738.773] (assumes normal distribution)


Result "org.apache.flink.benchmark.TwoInputBenchmark.twoInputOneIdleMapSink":
  4642.515 ±(99.9%) 192.741 ops/ms [Average]
  (min, avg, max) = (3219.666, 4642.515, 5885.845), stdev = 703.180
  CI (99.9%): [4449.774, 4835.257] (assumes normal distribution)


# Run complete. Total time: 00:32:36

Benchmark                                  Mode  Cnt     Score     Error   Units
TwoInputBenchmark.twoInputMapSink         thrpt  150  9520.452 ± 218.321  ops/ms
TwoInputBenchmark.twoInputOneIdleMapSink  thrpt  150  4642.515 ± 192.741  ops/ms
```

**Comment by pnowojski** [comment]
Thanks for the update :) I will probably merge it later this day (after a benchmark request run completes).

**Comment by pnowojski** [comment]
Thanks, merging. Lets monitor long term trends in those benchmarks.

**Pull request merged**